Docs: Integrating with VTEX Headless CMS (#1881)
vtex/faststore: apps/site/pages/docs/headless-cms-integration/3-adding-content-types-and-sections.mdx

---
title: '3. Structuring Content types and Sections'
sidebar_label: 'Integrating your storefront with VTEX Headless CMS'
---

import { Callout, Tabs, Tab } from 'nextra-theme-docs'

<header>
# Part 3: Structuring Content types and Sections to the Headless CMS
</header>

Now that you have created your `cms` folder and have an overall understanding of how you'll use it, let's learn how to structure the Content Type, Section, and Translation Key files so you can create any integrations you desire. 

---

## Step by step

### Step 1 - Automatically syncing your changes

1. Before we start creating more complex definitions of Content Types and Sections, go to the root of your FastStore project and run `yarn dev` command to enable the watch mode for the FastStore CLI.  _Remember to use a development workspace._

<Callout type="info" emoji="ℹ️">
This command will monitor any changes in the project, including the new section you have created, and automatically merge it with the default sections available in [`faststore/core`](/docs/getting-started/3-project-structure#faststorecore).
</Callout>

> 2. Open another terminal and run `faststore cms-sync` to automatically sync your changes in the cms folder with the VTEX Headless CMS app.

Reviewer: ```suggestion
2. Open another terminal and run `faststore cms-sync` to automatically sync your changes in the `cms` folder with the VTEX Headless CMS app.
```

> 2. Open the **VTEX Admin**, access **Storefront > Headless CMS** and keep it next to your code to see your changes while editing the `cms` folder. Notice that refreshing the Admin may be necessary to see your changes.

Reviewer: ```suggestion
2. Open the **VTEX Admin**, access **Storefront > Headless CMS** and keep it next to your code to see your changes while editing the `cms` folder. Refreshing the Admin may be necessary to see your changes.
```

### Step 2 - Adding Content Types to the VTEX Headless CMS

> We already know that the `content-types.json` file is an array of JSON objects that describes the types of pages available for customization at the VTEX Headless CMS. Therefore, to create new Content Types, we just need to declare a new JSON object for each of our Content Types in the `content-types.json` file. However, we still need to discover which props these objects expect. So let's get back to our previous example.

Reviewer: ```suggestion
The `content-types.json` file is an array of JSON objects that describes the types of pages available for customization at the VTEX Headless CMS. Therefore, to create new Content Types, we need to declare a new JSON object for each of our Content Types in the `content-types.json` file. However, we still need to discover which props these objects expect. So let's get back to our previous example.
```

                  },
                  "required": [
                    "slug",
                    "title",
                    "description"
                  ],
                  "properties": {
                    "slug": {
                      "title": "Path",
                      "type": "string",
                      "default": "/"
                    },
                    "title": {
                      "title": "Default page title",
                      "description": "Display this title when no other title is available",
                      "type": "string",
                      "default": "FastStore Starter"
                    },
                    "description": {
                      "title": "Meta tag description",
                      "type": "string",
                      "default": "A beautifully designed store"
                    },
                    "canonical": {
                      "title": "Canonical url for the page",
                      "type": "string"
                    }
                  }
                }
              }
            ]
          }
        ]
      }
    ]

   ```
  </Tab>

  <Tab>

>     <img src="https://vtexhelp.vtexassets.com/assets/docs/src/cms-custom-page___e603fdbc92945b4627c6c5576e169ea6.png" width="40%" />

Reviewer: Não entendi muito bem a relação dessa imagem de exemplo com o sample de código da outra tab

Author: É o resultado que aparece no Headless CMS do que declaramos no `content-types.json`.

  </Tab>
</Tabs>

Notice that, to declare a new Content Type, you must specify at least the following three parameters:

| Key | Description |
| --- | ------------|
| `id` | The identification key of a Content Type, written in Camel Case.   |
| `name` | The name presented at the VTEX Headless CMS app.  |

> | **(optional)** `configurationSchemaSets` | Creates a new tab on the page of that Content Type with advanced settings. In our example, the **Custom Page** includes a new tab called Settings with custom SEO settings. Check the [following section](#the-configurationschemasets-property) for more information. |

Reviewer: ```suggestion
| **(optional)** `configurationSchemaSets` | Creates a new tab on the page of that Content Type with advanced settings. In our example, the **Custom Page** includes a new tab called **Settings** with custom SEO settings. Check the [following section](#the-configurationschemasets-property) for more information. |
```

#### The `configurationSchemaSets` property

Back to our example, notice that the **Custom Page** Content Type has the `configurationSchemaSets` definition. 
The `configurationSchemaSets` prop creates a new tab at the VTEX Headless CMS for that Content Type with advanced settings.

For the **Custom Page**, a custom section named **Settings**, which allows editors to change the site metadata, was created.

![Configuration Schema Sets](https://vtexhelp.vtexassets.com/assets/docs/src/cms-settings-section___f6c8cfc11de6578e2a208ea1602ed127.png) 

Also, notice that the `configurationSchemaSets` object must contain a `name`, written in **camel case**, and `configurations`. 
In its turn, the `configurations` object must include a `name` for identification, written in camel case, and a `schema`, written in [JSON Schema v6](http://json-schema.org/) - a description language for creating forms. The `schema` defines the structure of the form used by editors to submit data.

<Callout type="info" emoji="ℹ️">
  For more information on how to write a `schema`, check the [`JSON Schema Reference`](https://json-schema.org/understanding-json-schema/index.html).
</Callout>

> 

Reviewer: ```suggestion
```

After editing the `cms/content-types.json` file, remember to save your changes and check them live by accessing the VTEX Admin at **Storefront > Headless CMS**. 

> Click on **Create New** and check the available Content Type options. Click on **Custom Page** to create a new **Custom Page** and check the Settings tab.

Reviewer: Fiquei com a sensação de que isso aqui poderia ser um passo a passo, já que tem várias instruções em um mesmo parágrafo.

> 

Reviewer: ```suggestion
```

### Step 3: Adding Sections to the VTEX Headless CMS

> Now, let's make some of our frontend components available for editing at the VTEX Headless CMS by writing the content schemas that represent them.

Reviewer: ```suggestion
Now, let's make some of our front-end components available for editing at the VTEX Headless CMS by writing the content schemas that represent them.
```

| Key           | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| :------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `title`       | The name that identifies your component in the VTEX Headles CMS app.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |

Reviewer: ```suggestion
| `title`       | The name that identifies your component in the VTEX Headless CMS app.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
```

Depending on the `type` of your schema, you may need to define particular fields related to your component structure. For example, for a schema of the `object` type, you'll need to determine `properties` that map key-value pairs. For a schema of the `array` type, you'll need to define the `items` of that array.

<Callout type="info" emoji="ℹ️">
  For more information on each property of a `schema`, check the [`JSON Schema Reference`](https://json-schema.org/understanding-json-schema/index.html).
</Callout>

#### Using widgets

When defining your Section `schema`, you can also use the [`uiSchema`](https://react-jsonschema-form.readthedocs.io/en/docs/api-reference/uiSchema/) along with [`widgets`](https://react-jsonschema-form.readthedocs.io/en/docs/usage/widgets/) to specify which UI widget should be used to render a given field of your schema. Common widgets are `draftjs-rich-text`, `image-uploader`, and `block-select`.

{/* TODO: Check the following example of the `draftjs-rich-text` component being used. */}

### Step 4 - Defining translation keys

> After creating our Section schemas and Content Types, 
> we will declare the Translation Keys in the `translation-keys.json` file. 

Reviewer: ```suggestion
After creating our Section schemas and Content Types, we will declare the Translation Keys in the `translation-keys.json` file. 
```

> Translation Keys are used to translate the form presented for editors to any locale you'd like.

Reviewer: Como separar traduções para cada locale? São arquivos diferentes? Ou dá para definir isso na própria key? Acho que vale explicar melhor aqui.

Author: Boa, Ju! Concordo. Só vou confirmar com o time se ainda faz sentido esse arquivo, pois estavam considerando deprecá-lo.

---
title: '3. Adding Content Types and Sections'
sidebar_label: 'Integrating your storefront with VTEX Headless CMS'
---

import { Callout, Tabs, Tab } from 'nextra-theme-docs'

<header>

> # Part 3: Adding Content Types and Sections to the VTEX Headless CMS

Author: Reminder: Change the doc name

</header>

Now that you have created your `cms` folder and have an overall understanding of how you'll use it, let's learn how to structure the Content Type, Section, and Translation Key files so you can create any integrations you desire. You'll also learn to use the `--watch` argument to automatically sync your changes with the VTEX Headless CMS app while developing your schemas.

---

## Step by step

### Step 1 - Automatically syncing your changes

Before we start creating more complex definitions of Content Types and Sections, go to the root of your FastStore project and run the following command to automatically sync your changes in the `cms` folder with the VTEX Headless CMS app. _Remember to use a development workspace._

```bash
vtex cms sync --watch

> ```

Author: change to `yarn dev`
